[MRG+2] default gamma='auto' in SVC

gxyd opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs

Fixes #8361 
Fixes #8535 
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.

Deprecates the default SVC gamma parameter value of "auto", which is calculated as 1 / n_features, and introduces "scale", which is calculated as 1 / (n_features * X.std()).


#### Any other comments?

I couldn't fix all the doctests since I'm not sure how to run doctests using pytest (I asked on gitter, though haven't received a response), otherwise using `make test` will more time to see which docs need to be fixed.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

neokt committed:
gamma=auto in SVC #8361

gxyd committed:
Merge branch 'master' into svc_gamma

Conflicts:
	sklearn/model_selection/_search.py
	sklearn/model_selection/tests/test_search.py

gxyd committed:
fix docs and add gamma='auto_deprecated' as default

gxyd committed:
fix

gxyd commented:
Currently it raises an error:
```
E           AssertionError: Estimator SVC should not change or mutate  the parameter gamma
             from auto_deprecated to auto during fit.
```

gxyd commented:
I can fix that issue I mention above (that is not a problem). I'v one query, that should we use `gamma='auto'` or `gamma='scale'` in docs? Also a lot of examples in documentation don't actually set `random_state` (one of the points Raghav raised here https://github.com/scikit-learn/scikit-learn/pull/8535#discussion_r105567684), do I need to set `random_state`? Atleast for the example I touch in the documentation.

gxyd committed:
revert change to change self.gamma

gxyd commented:
@amueller can you give a look over the PR, though I've few questions as well that I've mentioned.

gxyd committed:
fix docs

gxyd commented:
Locally tests pass. I am guessing that travis uses scipy version 0.13.3, which doesn't have the `power` method for `lil_matrix`, hence the current error on travis. Travis is based on Ubuntu 14.04 (which has scipy 0.13.3 upstream) and that is causing the problem.

gxyd commented on sklearn/svm/base.py:
```diff
@@ -169,7 +169,19 @@ def fit(self, X, y, sample_weight=None):
-        if self.gamma == 'auto':
+        if self.gamma == 'scale':
+            if isinstance(X, sp.spmatrix):
+                X_std = np.sqrt(X.power(2).mean() - (X.mean())**2)
```
I've been able to figure out how to do this without the use of `X.power`.

gxyd replied:
I could simply use `X.toarray()**2`, but I don't think that is efficient.

lesteve replied:
Maybe `X.multiply(X)`?

amueller replied:
Why can there be a ``lil`` matrix here? We called ``check_X_y`` above with ``accept_sparse='csr'``.

amueller reviewed (commented):
Can we maybe not deprecate ``auto`` and just change the default? not sure, though....

gxyd committed:
use X.multiply(X) for csr_matrix

gxyd committed:
fix pep8

gxyd commented:
@amueller 
> Can we maybe not deprecate auto and just change the default? not sure, though....

I think that is the intention here. That is to replace the default value not to actually remove 'auto' from possible gamma values.
Does the usage of `gamma=auto_deprecate` sounds misleading here? Or better simply use `auto_default`? Though @jnothman suggested that here https://github.com/scikit-learn/scikit-learn/pull/8535#discussion_r104302523 . Also I think the warning message should be made more clear. WDYT?

amueller commented on sklearn/svm/classes.py:
```diff
@@ -448,20 +448,27 @@ class SVC(BaseSVC):
-        If gamma is 'auto' then 1/n_features will be used instead.
+        If gamma is 'auto' then 1/n_features will be used.
+        If gamma is 'scale' then 1/(n_features * X.std()) will be used.
+        The current default 'auto' is deprecated in version 0.20 and will
```
I guess the formulation here was what led me to think that ``'auto'`` is deprecated, i.e. will be removed. I would rather say that "The default value will change to 'scale' in 0.22".  I'm not sure about a good way to say when this decision was made. I think we had "This warning was introduced in version X" somewhere.

jnothman commented:
It's tempting to just change the meaning of auto, isn't it? But as it
affects the model, we probably shouldn't...

gxyd commented:
I can't say if it is tempting, since I don't know the reason for making 
this change. I mean, why is using gamma=1 / (n_features * X.std()) is 
more apt choice? Can you please refer me to material explaining the 
reason for why it performs better than gamma=1 / n_features.

And if it does lead to improvement in model, then why not? And reading 
from the discussion on the issue @amueller (at that time) seemed quite 
convinced of the benefits of this change in gamma value.

amueller commented:
There's probably no literature on why this would be good. But the ``1/n_features`` heuristic only makes sense when ``X`` is scaled. If you scaled ``X``, there will be no change. If you didn't scale ``X``, it'll "fix" that somewhat. Clearly, gamma is inversely related to scaling of the data; if you scale X by 10 you need to scale gamma by 1/10.
Why not change it? Because that'll change the behavior of existing code and people will suddenly get different results.

gxyd commented:
I had a look over "A Practical Guide to Support Vector Classification" (by Chih-Wei Hsu, Chih-Chung Chang, and Chih-Jen Lin), and it says that
> Scaling before applying SVM is very important. Part 2 of Sarle’s Neural Networks
FAQ Sarle (1997) explains the importance of this and most of considerations also ap-
ply to SVM.

So considering that, it would imply that in most of the cases scaling is mostly needed.

> Why not change it? Because that'll change the behavior of existing code and people will suddenly get different results.

Yes, I totally understand this part. "Changing" of things is one of the things to upset users.

Though I'll take your final word, if you still consider that the change is worth it.

amueller commented:
> So considering that, it would imply that in most of the cases scaling is mostly needed.

Yes, but people don't do it. Basically the new default helps people who forget to scale their data.

gxyd commented:
That is right. Should I close the PR then? May be the corresponding issue as well? (I can't close the issue, I didn't open).

amueller commented:
What? No. Please go ahead with this. I would do the full deprecation cycle as you did.

amueller commented:
I would just clarify the wording on what exactly changes.

gxyd committed:
use 'auto_default' instead of 'auto_deprecated'

gxyd commented:
I don't understand what the problem with CircleCI tests are. Can you 
help me out?

gxyd commented:
I don't think the error is related to the PR. It raises
```
CondaValueError: invalid package specification: python=
```
So possibly something wrong with the build configuration itself.

gxyd commented:
@amueller any suggestion for the failing tests?

amueller commented:
Ignore it ;)

amueller commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -975,3 +975,21 @@ def test_ovr_decision_function():
+def test_gamma_scale():
+    X, y = [[0.0], [1.0]], [0, 1]
+
+    clf = svm.SVC(gamma='scale').fit(X, y)
```
Can you check that it doesn't warn?

amueller commented on sklearn/svm/classes.py:
```diff
@@ -575,7 +582,7 @@ class SVC(BaseSVC):
     _impl = 'c_svc'
 
-    def __init__(self, C=1.0, kernel='rbf', degree=3, gamma='auto',
+    def __init__(self, C=1.0, kernel='rbf', degree=3, gamma='auto_default',
```
Hm I think I liked ``auto_deprecated`` better? bikeshedding a bit, though.

gxyd replied:
Na, in some sense `auto_deprecated` is better, since it gives a clear indication to change back this variable's name in the future.

Do I need to document it in the docstring?

jnothman replied:
It certainly can't hurt

amueller commented:
Rest looks good. Given all the changes you made, I assume you checked that no new deprecation warnings are raised in the tests?
I was first a bit concerned about the deprecation warnings, but really, no-one should seriously use SVC with default values.

gxyd commented:
@jnothman how do I run just the doctests?

jnothman commented:
just run pytest on those files..?

gxyd commented:
I mean how do I only run the doctests? i.e. testing those files which 
are in doc/ directory, and not test those in sklearn/ directory.

gxyd commented:
It takes a lot of time if I run `make test` (since it also runs tests 
for sklearn/ directory).

gxyd commented:
I understand, I thought of that as a question. Thanks :)

gxyd committed:
fix warning in docs

gxyd committed:
don't raise warnings in sklearn

gxyd committed:
fix warning

gxyd commented:
@amueller 
> Given all the changes you made, I assume you checked that no new deprecation warnings are raised in the tests?

No, I hadn't taken care of that. I have tried to address that in the last few commits. Though I'm facing one problem that, what should we do in case of `X.std() == 0`, since then it will probably raise a `RunTimeWarning` as currently happens in here https://github.com/gxyd/scikit-learn/blob/a41459a8c718c781950db570c0257a0490059e70/sklearn/ensemble/tests/test_weight_boosting.py#L304 (instead of raising `ValueError` rather it raises `RunTimeError`).

amueller commented:
How about setting gamma just to 1 there?

gxyd commented:
That would work good. And do we intend to change value of `gamma` (from 
`auto` to `scale`) just for the `SVC` , what about `NuSVC`, `SVR`, 
`NuSVR`, `OneClassSVM` ... as well?

amueller commented:
all of them :-/

gxyd committed:
change gamma='auto' -> gamma='auto_deprecated'

gxyd committed:
fix a few test warnings

gxyd committed:
some more changes

gxyd committed:
fix warning

gxyd committed:
fix pycodestyle errors

gxyd commented:
@amueller Currently it fails travis with with error:
```bash
Running flake8 on the diff in the range 61e722aa12..c6a9325872 (17 commit(s)):
--------------------------------------------------------------------------------
sklearn/svm/tests/test_svm.py:514:45: F821 undefined name 'unicode'
        clf = svm.SVC(gamma='scale', kernel=unicode('linear'))
```

Why does it reach this block of code with python3.5? I think `six.PY2` will be false for python3.5, even then it raises this error.

gxyd commented:
@jnothman can you suggest any possible explanation of failing tests in the comment above? I still can't understand why that should be failing.

gxyd commented:
Ping @jnothman

jnothman commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -814,6 +814,12 @@ def test_grid_search_cv_results():
+    grid_search = GridSearchCV(SVC(gamma="scale"), cv=n_splits, iid=False,
+                               param_grid=params)
+    grid_search.fit(X, y)
+    grid_search_iid = GridSearchCV(SVC(gamma="scale"), cv=n_splits, iid=True,
```
You don't need gamma='scale' here and below where gamma is being searched over: the SVC will never be fit with gamma='auto_deprecated'

gxyd replied:
I don't understand as to why this changes isn't needed. If we don't use `gamma='scale'` here explicitly, then this will raise a warning, which isn't desired considering https://github.com/scikit-learn/scikit-learn/pull/10331#issuecomment-353168653.

jnothman commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -975,3 +980,22 @@ def test_ovr_decision_function():
+def test_gamma_auto():
+    X, y = [[0.0], [1.0]], [0, 1]
+
+    msg = ("The default gamma parameter value 'auto', calculated as "
```
I think this would be clearer as "The default value of gamma will change from 'auto' to 'scale' in version X to account better for unscaled features". The user can read the docs.

gxyd replied:
Done.

jnothman replied:
Actually it might be good to add "Set gamma explicitly to 'auto' or 'scale' to avoid this warning."

jnothman reviewed (approved):
I don't think you need to worry about the flake8 error.

gxyd committed:
better error message

jnothman commented:
Other reviewers should disregard the flake8 error.

gxyd committed:
improve error message, document 'auto_deprecated'

gxyd commented on sklearn/svm/classes.py:
```diff
@@ -620,7 +629,16 @@ class NuSVC(BaseSVC):
+        Current default is 'auto' which uses 1 / n_features,
+        if gamma='scale' is passed then it uses 1 / (n_features * X.std())
+        features. The current default of gamma, 'auto', will change to 'scale'
+        in version 0.22. 'auto_deprecated', a deprecated version of 'auto' is
```
I've documented 'auto_deprecated' here. Does this look okay (place and wording)?

gxyd committed:
add entry in whats_new

gxyd commented:
Let me know if I need to resolve the merge conflicts (possibly due to *change log* file changes) myself.

gxyd commented:
Oh, okay these are due to svm/tests/test_svm, I'll fix those.

gxyd committed:
Merge branch 'master' into svc_gamma

Conflicts:
	sklearn/svm/tests/test_svm.py

gxyd commented:
Resolved the conflicts here, probably that should do okay.

qinhanmin2014 commented on doc/whats_new/v0.20.rst:
```diff
@@ -155,6 +155,12 @@ Linear, kernelized and related models
+  :class:`svm.NuSVC`, :class:`svm.SVR`, :class:`NuSVR`, :class:`OneClassSVM`
+  will change from ``'auto'`` to ``'scale'`` in version 0.22 to account
+  better for unscaled features. :issue:`8361` by :user:`Gaurav Dhingra <gxyd>`
+  and :user:`Ting Neo <neokt>`.
```
I think you are in the wrong section (there's an API changes summary section) #

gxyd committed:
move what's new entry to more appropriate section

gxyd commented:
Thanks for the review :) I've moved the what's new entry to an appropriate section.

> It makes many examples more complicated (for about 2 years I suppose?). Default parameters in the examples is friendly to newcomers. I don't want to see, e.g., someone learning model persistence wondering why we add gamma='scale' in SVC.

I agree on that for beginners having something as this can be scary (I'm a beginner too).

But, I believe we should have defaults that give the best possible results, even though (as you said) "suppose anyone who use svm seriously should tune the parameter by themselves", I would still prefer having good defaults above this, even if it means having (~2 years) of raising warning when no `gamma` is passed manually.

jnothman reviewed (commented):
One thing I failed to notice: we should not be warning if kernel in ('linear', 'precomputed') as these don't use gamma. This will slightly limit the damage.

gxyd committed:
revert gamma='scale' change from kernel in ('linear', 'precomputed')

gxyd committed:
revert few more changes

gxyd commented:
I'm sorry that I didn't notice this earlier, I should had. I've made the 
changes, probably all of them (including reverting the tests).

Should we add a test, depicting that indeed /no warning/ is raised for 
/kernel/ in ('linear', 'precomputed')?

jnothman commented:
Sure, do so.

gxyd committed:
add tests for kernel='linear', 'precomputed' not raising error

gxyd committed:
fix flake8 error

gxyd commented:
The travis is again on a backlog I guess.

gxyd commented:
Does everything look okay now?

qinhanmin2014 commented:
I've restarted travis for you @gxyd . Thanks for the great work :)
I'll wait for others to give the +2. I understand although users should scale their data and tune the parameters, it's still good to provide a better default value, but I'm still a bit reluctant to make so many examples complicated (including many basic parts, e.g., tutorial, pipeline).

gxyd commented:
Thank for very much. I agree with you on somethings, but I'm still 
holding onto my opinion. Personally, may be I'll need some more time 
with sklearn community and machine learning opinions to better 
appreciate your concerns.

Thanks again for the review.

jnothman commented:
Well it might be worth leaving ourselves a note to change things back in examples after deprecation ends.

gxyd committed:
add a Note to revert examples, when deprecation ends

gxyd commented:
I've added a note, I think that'll suffice.

gxyd committed:
reword NOTE

jnothman reviewed (commented):
@qinhanmin2014, I agree it's not ideal to have to make such a compromise. Especially when there's a little dispute as to whether or not we should be helping the user with this scaling at all (but @amueller seems to have his ear on the user-base pulse better than anyone else and asked for this to be resolved, so I say we trust him there). Does the idea that we make sure to clean up the examples in two years' time placate you at all?

qinhanmin2014 commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -814,6 +814,12 @@ def test_grid_search_cv_results():
     n_grid_points = 6
     params = [dict(kernel=['rbf', ], C=[1, 10], gamma=[0.1, 1]),
               dict(kernel=['poly', ], degree=[1, 2])]
+    grid_search = GridSearchCV(SVC(gamma="scale"), cv=n_splits, iid=False,
```
Why do you need to add these lines? Seems duplicate?
https://github.com/gxyd/scikit-learn/blob/890ce677065a4ef90a5c3771e096a5876c46856b/sklearn/model_selection/tests/test_search.py#L836-L839

gxyd replied:
Oh, I see. This isn't intentional, good catch though :) . What happened is, when I merged the master branch (2nd commit in git history), after basing my commit of @neokt (@neokt initially started working on it), I didn't resolve the commits cleanly, since there were a few lines in master that had been removed later on.

So, all in all it was a case of me not cleanly resolving merge conflicts.

qinhanmin2014 commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -863,6 +870,16 @@ def test_random_search_cv_results():
 
     n_splits = 3
     n_search_iter = 30
+    params = dict(C=expon(scale=10), gamma=expon(scale=0.1))
```
Why do you need to add these lines? Seems duplicate?
https://github.com/gxyd/scikit-learn/blob/890ce677065a4ef90a5c3771e096a5876c46856b/sklearn/model_selection/tests/test_search.py#L897-L900

gxyd replied:
probably the same case here as well (not resolving the merge conflicts cleanly).

qinhanmin2014 commented on sklearn/svm/classes.py:
```diff
@@ -446,20 +446,29 @@ class SVC(BaseSVC):
+
+        Current default is 'auto' which uses 1 / n_features,
+        if gamma='scale' is passed then it uses 1 / (n_features * X.std())
+        features. The current default of gamma, 'auto', will change to 'scale'
```
what do you mean by 1 / (n_features * X.std()) features?

gxyd replied:
better replace 'features' here with 'as the kernel coefficient'?

gxyd replied:
Or may be simply, 'features' -> 'as a value of gamma'. This one I think is better.

qinhanmin2014 reviewed (commented):
> but @amueller seems to have his ear on the user-base pulse better than anyone else and asked for this to be resolved, so I say we trust him there

@jnothman Agree :) I'll give my LGTM after checking the code.

gxyd committed:
Merge branch 'master' into svc_gamma

gxyd committed:
fix messed up addition while resolving merge conflicts

gxyd committed:
small change

qinhanmin2014 reviewed (approved):
LGTM ping @jnothman

gxyd commented on sklearn/svm/base.py:
```diff
@@ -168,7 +168,28 @@ def fit(self, X, y, sample_weight=None):
+                              "'auto' to 'scale' in version 0.22 to account "
+                              "better for unscaled features. Set gamma "
+                              "explicitly to 'auto' or 'scale' to avoid this "
+                              "warning.", DeprecationWarning)
```
I think `FutureWarning` suites here better?

jnothman commented:
Another thing. Should we not raise the warning if X.std() is approximately 1?

gxyd commented:
Which warning? when X.std() is approximately 1 (1 right? not 0) ?

jnothman commented:
yes, 1.

We don't need to warn that the default will change as the new value would
have identical effect for that training set

gxyd committed:
don't raise warning when X_std is approximately 1

Since when X_std ~= 1, then both gamma='auto' or gamma='scale', will
have the same effect

gxyd commented:
Yes, ofcourse, good observation indeed. I've fixed it and added a test 
case for it.

jnothman commented:
Are there more examples where we didn't need to change?? ;)

On 24 January 2018 at 19:18, Gaurav Dhingra <notifications@github.com>
wrote:

> Yes, ofcourse, good observation indeed. I've fixed it and added a test
> case for it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10331#issuecomment-360054233>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz60sJHi-__F-Sl_lE8myyiYGnOycwks5tNudmgaJpZM4RESD1>
> .
>

gxyd committed:
add comment about test case

gxyd committed:
undo using gamma explicitly since, X_std is approx. ~=1

gxyd committed:
Merge branch 'master' into svc_gamma

gxyd commented on doc/modules/cross_validation.rst:
```diff
@@ -165,7 +165,8 @@ validation iterator instead, for instance::
 
       >>> from sklearn.pipeline import make_pipeline
-      >>> clf = make_pipeline(preprocessing.StandardScaler(), svm.SVC(C=1))
+      >>> clf = make_pipeline(preprocessing.StandardScaler(),
```
I'll undo this, just puts the text on next line currently.

gxyd commented:
I'll have a one more good look over each change, manually. And see if any more can be reverted.

gxyd committed:
revert gamma='scale' (part 1)

gxyd committed:
revert, unrelated change

gxyd committed:
revert complete

gxyd commented:
Now, I don't think any more changes can be reverted.

jnothman reviewed (approved):
Thanks for the meticulous work. I've tried to reduce the unnecessary warnings and it's still touching a lot. A further way to avoid this would be to scale the data in examples, but I'm not sure we want to do that.

I'd like to see if @amueller could comment on how this looks, but otherwise we should probably merge

qinhanmin2014 commented on sklearn/ensemble/tests/test_bagging.py:
```diff
@@ -115,15 +115,17 @@ def fit(self, X, y):
                 # Trained on sparse format
                 sparse_classifier = BaggingClassifier(
-                    base_estimator=CustomSVC(decision_function_shape='ovr'),
+                    base_estimator=CustomSVC(gamma='auto',
```
Any specific reason using gamma='auto' here? It will make us unable to remove it after 0.22.

gxyd replied:
When I set this explicitly to `gamma='scale'`, it was raising a error:
```python
================================================================== FAILURES ===================================================================
_________________________________________________________ test_sparse_classification __________________________________________________________

    def test_sparse_classification():
        # Check classification for various parameter settings on sparse input.
    
        class CustomSVC(SVC):
            """SVC variant that records the nature of the training set"""
    
            def fit(self, X, y):
                super(CustomSVC, self).fit(X, y)
                self.data_type_ = type(X)
                return self
    
        rng = check_random_state(0)
        X_train, X_test, y_train, y_test = train_test_split(iris.data,
                                                            iris.target,
                                                            random_state=rng)
        parameter_sets = [
            {"max_samples": 0.5,
             "max_features": 2,
             "bootstrap": True,
             "bootstrap_features": True},
            {"max_samples": 1.0,
             "max_features": 4,
             "bootstrap": True,
             "bootstrap_features": True},
            {"max_features": 2,
             "bootstrap": False,
             "bootstrap_features": True},
            {"max_samples": 0.5,
             "bootstrap": True,
             "bootstrap_features": False},
        ]
    
        for sparse_format in [csc_matrix, csr_matrix]:
            X_train_sparse = sparse_format(X_train)
            X_test_sparse = sparse_format(X_test)
            for params in parameter_sets:
                for f in ['predict', 'predict_proba', 'predict_log_proba', 'decision_function']:
                    # Trained on sparse format
                    sparse_classifier = BaggingClassifier(
                        base_estimator=CustomSVC(gamma='scale',
                                                 decision_function_shape='ovr'),
                        random_state=1,
                        **params
                    ).fit(X_train_sparse, y_train)
                    sparse_results = getattr(sparse_classifier, f)(X_test_sparse)
    
                    # Trained on dense format
                    dense_classifier = BaggingClassifier(
                        base_estimator=CustomSVC(gamma='scale',
                                                 decision_function_shape='ovr'),
                        random_state=1,
                        **params
                    ).fit(X_train, y_train)
                    dense_results = getattr(dense_classifier, f)(X_test)
>                   assert_array_equal(sparse_results, dense_results)
E                   AssertionError: 
E                   Arrays are not equal
E                   
E                   (mismatch 36.8421052632%)
E                    x: array([[-0.288395,  0.936657,  2.351738],
E                          [-0.351292,  1.731076,  1.620216],
E                          [-0.369671,  1.825747,  1.543924],...
E                    y: array([[-0.288395,  0.936657,  2.351738],
E                          [-0.351292,  1.731076,  1.620216],
E                          [-0.369671,  1.825747,  1.543924],...

```

Though I didn't investigate it.

jnothman commented:
> It will make us unable to remove it after 0.22.

Why?

gxyd commented:
> A further way to avoid this would be to scale the data in examples, but I'm not sure we want to do that.

That'll still either need to explicitly set `gamma='auto'` or raise a *new* warning in the test suite, neither of which is desired I think.

jnothman commented:
why would they need to explicitly set gmma to auto if std=1? and I'm
referring to examples, not tests

On 26 Jan 2018 4:10 pm, "Gaurav Dhingra" <notifications@github.com> wrote:

> A further way to avoid this would be to scale the data in examples, but
> I'm not sure we want to do that.
>
> That'll still either need to explicitly set gamma='auto' or raise a *new*
> warning in the test suite, neither of which is desired I think.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10331#issuecomment-360686026>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6xWkWpt-ZMVHXyhNcDd72xnFZDjCks5tOV43gaJpZM4RESD1>
> .
>

gxyd commented:
I misunderstood. I thought you were referring to any case (not just for 
X.std()=1).

As for the comment by @qinhanmin2014 , should I bother to investigate as 
to why that test case fails with gamma='scale' but not with 
gamma='auto'. Would it be necessary to investigate that?

qinhanmin2014 commented:
>  It will make us unable to remove it after 0.22.
>  Why?

@jnothman 
The above test is the only place we add `gamma='auto'` (in other places we add `gamma='scale'`). This test seems irrelevant to gamma so I think we need to remove the assignment of gamma after 0.22. Setting `gamma='auto'` will make us unable to do that.

qinhanmin2014 commented:
> As for the comment by @qinhanmin2014 , should I bother to investigate as to why that test case fails with gamma='scale' but not with gamma='auto'. Would it be necessary to investigate that?

@gxyd I think it's not harmful to investigate since I still think this test is irrelevant to gamma. At a glance, I think replacing `assert_array_equal` with `assert_almost_equal` might solve the problem? It seems unreasonable to do float comparison with `assert_array_equal`.

gxyd commented:
> At a glance, I think replacing assert_array_equal with assert_almost_equal might solve the problem?

It does indeed :). I wasn't expecting that after seeing mismatch 36.8421052632% (didn't know the way this percentage is calculated, but I do now https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.testing.assert_array_almost_equal.html)

For the above discussion, anyway we can now explicitly set `gamma='scale'` instead (`assert_array_almost_equal` seems to working just fine).

gxyd committed:
use assert_array_almost_equal instead of assert_array_equal

qinhanmin2014 commented on doc/modules/metrics.rst:
```diff
@@ -177,7 +177,7 @@ It can be computed using :func:`chi2_kernel` and then passed to an
 It can also be directly used as the ``kernel`` argument::
 
-    >>> svm = SVC(kernel=chi2_kernel).fit(X, y)
+    >>> svm = SVC(gamma='scale', kernel=chi2_kernel).fit(X, y)
```
@gxyd also ping @jnothman 
I wonder whether it makes sense to pass gamma='scale' here, See below:
```python
X = [[0, 1], [1, 0], [.2, .8], [.7, .3]]
y = [0, 1, 0, 1]
svm = SVC(gamma=0.5, kernel=chi2_kernel, probability=True).fit(X, y)
test1 = svm.predict_proba(X)
svm = SVC(gamma=1, kernel=chi2_kernel, probability=True).fit(X, y)
test2 = svm.predict_proba(X)
assert_array_almost_equal(test1, test2) # True, gamma is ignored and set to default(1.0)

K = chi2_kernel(X, gamma=0.5)
svm = SVC(kernel='precomputed', probability=True).fit(K, y)
test3 = svm.predict_proba(K)
K = chi2_kernel(X, gamma=1)
svm = SVC(kernel='precomputed', probability=True).fit(K, y)
test4 = svm.predict_proba(K)
assert_array_almost_equal(test1, test3) # False
assert_array_almost_equal(test1, test4) # True
```
In this case, we'll get unnecessary warnings since gamma is ignored? Or maybe it's a bug? (Or sorry if I'm wrong)

gxyd replied:
I don't think this should be considered a bug, since the documentation of gamma says:
```
gamma : float, optional (default=’auto’)

    Kernel coefficient for ‘rbf’, ‘poly’ and ‘sigmoid’. If gamma is ‘auto’ then 1/n_features will be used instead.
```

It doesn't say, that 'gamma' will be used for a 'callable' or 'precompute'. Unless we consider this to be some problem in the documentation, I see no problem with current results of SVC neglecting gamma, when a callable is passed.

So, implying that it isn't meaningful for a callable kernel or for kernel='precompute'.

qinhanmin2014 replied:
> I don't think this should be considered a bug

Agree. We have other ways (e.g., functools.partial) to specify parameters for a callable. So correct the misleading example and avoid unnecesssary warnings seem enough from my side.

gxyd replied:
Yes, that's how I tested locally as well. I wasn't aware of `functools.partial` until I worked on my other PR (regarding jaccard_similarity_score).

gxyd committed:
don't raise warning for a callable

gxyd commented:
I've treated a callable (which includes 'chi2_kernel') in a similar way as kernel='linear' or kernel='percompute', since I don't think gamma makes sense in case of a callable either.

@qinhanmin2014 thank you very much for raising this point, it allowed me to step through a debugger to see the code :) . The code snippet you provided is spot-on.

qinhanmin2014 reviewed (approved):
Thanks @gxyd I'm now much happier to get this in :) ping @jnothman @amueller

lesteve commented on sklearn/svm/classes.py:
```diff
@@ -446,20 +446,29 @@ class SVC(BaseSVC):
+        'auto' is used as a default indicating that no explicit value of gamma
+        was passed.
+
+        .. versionchanged:: 0.20
```
Not quite sure about what to put as versionchanged here. A bit confusing at first but I guess "versionchanged:: 0.20" is fine ...

lesteve replied:
Still unsure about this one. Either we:
* do not add anything now and try to remember to add a versionchanged in 0.22 to say that the default has changed. 
* use `versionchanged:: 0.22` and write it as a message from our future selves i.e. "default has changed in 0.22 from 'auto' to 'scale'.

I think I favour option 1. less confusing.

qinhanmin2014 replied:
Sorry I somehow missed the comment here. I'm fine with removing versionchanged
(1) There's enough information in parameter description
(2) The change will actually occurs in 0.22, not 0.20.

lesteve commented on sklearn/svm/base.py:
```diff
@@ -168,8 +168,32 @@ def fit(self, X, y, sample_weight=None):
+                                  "from 'auto' to 'scale' in version 0.22 to "
+                                  "account better for unscaled features. Set "
+                                  "gamma explicitly to 'auto' or 'scale' to "
+                                  "avoid this warning.", DeprecationWarning)
```
Maybe a FutureWarning rather than a DeprecationWarning? We are not really deprecating anything, but warning about changing the default in the future.

gxyd replied:
I suggested that here https://github.com/scikit-learn/scikit-learn/pull/10331#discussion_r163312773 but I don't remember why then I thought otherwise.

@jnothman opinion?

lesteve commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -230,22 +231,22 @@ def test_svr_errors():
 def test_oneclass():
     # Test OneClassSVM
-    clf = svm.OneClassSVM()
+    clf = svm.OneClassSVM(gamma='scale')
```
That should be gamma='auto' in order not to change the result, shouldn't it?

gxyd replied:
> in order not to change the result

But I don't think that's the focus of this test is. It is to actually test the default `OneClassSVM`, which will eventually have (in version 0.22) `gamma='scale'` as default.

Infact at most of the places we have used `gamma='scale'` and not `gamma='auto'`.

lesteve replied:
> Infact at most of the places we have used gamma='scale' and not gamma='auto'.

I saw that and I am not sure I understand why I have to admit ... is it to test `gamma='scale'` because it is the upcoming default, is it because at least docs should promote best practices, is it for some other reason that I am missing?

qinhanmin2014 replied:
The logic here is to add `gamma='scala'` in the tests now, and to remove them when default gamma is set to `scale` in the future. I agree with you that it might be better to use `gamma=auto` in this test to avoid modifying the result, thus to avoid protential test failures.
Since I don't think it will make too much difference, both solutions LGTM from my side.

lesteve replied:
I think it is fine to leave it like this. At least the future default is well covered by tests.

qinhanmin2014 replied:
Fine, let's leave it like this :)

lesteve reviewed (commented):
A few comments, but otherwise LGTM.

lesteve commented on sklearn/svm/base.py:
```diff
@@ -168,8 +168,32 @@ def fit(self, X, y, sample_weight=None):
+                    self._gamma = 1.0
+            else:
+                if (self.gamma == 'auto_deprecated' and
+                        self.kernel not in ('linear', 'precomputed') and
```
This got lost somewho. I would rewrite this a bit for better readability:
```py
kernel_uses_gamma = self.kernel not in ('linear', 'precomputed') and not callable(self.kernel)
if (kernel_uses_gamma and not np.isclose(X_std, 1.0)):
    ...
```

gxyd replied:
That would look neat.

qinhanmin2014 commented:
ping @gxyd Could you please address the above comments? I think it's very closed from being merged.

gxyd committed:
Merge branch 'master' into hello_world

Conflicts:
	doc/whats_new/v0.20.rst

gxyd commented on sklearn/svm/base.py:
```diff
@@ -168,8 +168,32 @@ def fit(self, X, y, sample_weight=None):
 
-        if self.gamma == 'auto':
-            self._gamma = 1.0 / X.shape[1]
+        if self.gamma in ('scale', 'auto', 'auto_deprecated'):
```
That's right. I think that can be moved as a separate condition. Will fix in the next commit.

gxyd committed:
address lesteve's comments

qinhanmin2014 commented:
@gxyd Thanks for the update.
(1) I also vote for FutureWarning (since we are not going to deprecate 'auto'?) If you can't find comments from other core devs which oppose FutureWarning, please update accordingly.
(2) Since we're not going to deprecate 'auto', I agree with lesteve that we should use gamme='auto' if gamme='scale' changes the result of the test, in order to avoid protential test failures.
Already +4 from jnothman, amueller, lesteve and me so I'll merge once you address these two comments.

qinhanmin2014 committed:
DeprecationWarning -> FutureWarning

qinhanmin2014 commented:
I've changed DeprecationWarning to FutureWarning, will merge when green.

qinhanmin2014 committed:
remove versionchanged

qinhanmin2014 reviewed (approved):
LGTM. Thanks @gxyd :)

gxyd commented:
Thanks @jnothman , @lesteve , @amueller and @qinhanmin2014 (making those last changes) for reivew.

mgelbart commented:
Hi all, this is my first time posting here; not sure if this is an appropriate post or if people look at already-merged PR threads. However, I am wondering about the following:

> There's probably no literature on why this would be good. But the `1/n_features` heuristic only makes sense when `X` is scaled. If you scaled `X`, there will be no change. If you didn't scale `X`, it'll "fix" that somewhat. Clearly, gamma is inversely related to scaling of the data; if you scale X by 10 you need to scale gamma by 1/10.

It seems to me that if you scale `X` by 10 you need to scale `gamma` by 1/100, not 1/10. If you look at the definition of the RBF kernel [here](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.rbf_kernel.html), you'll see that the "units" of `gamma` are 1/x^2, not 1/x. I also tested this empirically: scaling `X` by 10 and scaling `gamma` by 1/100 gives the same result as the original, whereas scaling `X` by 10 and scaling `gamma` by 1/10 gives a different result. If I'm not mistaken, perhaps a better default would be `1 / (n_features * X.var())` instead of `1 / (n_features * X.std())`?

Related issues: #8361, #10331

amueller commented:
@mgelbart I'm afraid you might be right and I have no idea how we overlooked that. Can you please open a new issue? This will be kinda hard to fix unfortunately :(

mgelbart commented:
@amueller just opened it (#12741).
